core/src/main/scala/filodb.core/memstore/IndexMetadataStore.scala

```diff
@@ -86,6 +91,249 @@ trait IndexMetadataStore {
+   * starting with most to least significant bytes
+   *
+   * In case the snap file does not exists, magic is not as expected or the file is corrupt, RebuildIndex will be
+   * triggered
+   *
+   * @param datasetRef The dataset ref
+   * @param shard the shard id of index
+   * @return a tuple of state and the option of time at which the state was recorded
+   */
+
+  override def currentState(datasetRef: DatasetRef, shard: Int): (IndexState.Value, Option[Long]) = {
+    val snapFile = new File(snapFilePathFormat.format(datasetRef.dataset, shard))
+    if (!snapFile.exists()) {
+      logger.info("Snap file {} does not exist , triggering rebuild", snapFile.toString)
+      (IndexState.Empty, None)
+    } else {
+      Using(new FileInputStream(snapFile)) { fin =>
+        getSnapTime(snapFile.toString, fin) match {
+          case Success(snapTime)    =>
+                                        logger.info("Found lastSyncTime={} for dataset={} and shard={}",
+                                          snapTime, datasetRef.toString, shard)
+                                        (IndexState.Synced, Some(snapTime))
+          case Failure(e)           =>  throw e           // Failure to read the snap file in currentState is fatal
+        }
+      }.get
+    }
+  }
+
+  /**
+   *  Gets the snapTime from the snap file
+   *
+   * @param snapFile   the snapFile
+   * @param fin        The input stream object for the file
+   * @return           Long snap time if file is valid
+   */
+  private def getSnapTime(snapFile: String, fin: FileInputStream): Try[Long] = {
+        val magicByte = fin.read()
+        if (magicByte != FileSystemBasedIndexMetadataStore.snapFileV1Magic) {
+          // if first byte is not the expected magicByte then the file is not as expected and possibly garbled
+          logger.warn(" Snap {} exist and possibly corrupt, this will trigger a rebuild", snapFile.toString)
+          Failure(new IllegalStateException("Invalid magic bytes in snap file"))
+        } else {
+          // If the first byte is magicByte read just the next 8 bytes
+          val bytes = Array[Byte](0, 0, 0, 0, 0, 0, 0, 0)
+          if (fin.read(bytes) != 8) {
+            // Not able to read 8 bytes, the file is truncated and we will trigger a rebuild
+            logger.warn("Snap file {} with right magic bytes but less than 8 for time," +
+              " this will trigger a rebuild of index", snapFile.toString)
+            Failure(new IllegalStateException("Snap file truncated"))
+          } else {
+            val lastSyncTime = ByteBuffer.wrap(bytes).getLong()
+            Success(lastSyncTime)
+          }
+        }
+   }
+  /**
+   *  Updates the state of the index, the snap file will be updated with the correct timestamp in case the state
+   *  is Synced, current implementation of updateState ignores other index states. Also when the index is updated
```

Comment:
updateState does some checks/deletes etc on state TriggerBuild and Empty ?

Reply:
Good catch.